src/arch/powerpc64.rs

```diff
@@ -0,0 +1,699 @@
+//! callgraph profiler. These work by following a linked list of frame records
+//! starting from the FP register. Each record consists of 2 words: a pointer
+//! to the previous frame (aka the previous FP value) and the return address
+//! for this frame (aka the saved LR value).
+//!
+//! To support these tools, we also generate a valid stack frame record when
+//! switching into a coroutine. This works by treating the parent link at the
+//! root of the stack as a frame record which points to the top of the parent
+//! stack. The top of the parent stack contains the saved FP and LR values in
+//! the correct format for a frame record, which allows unwinding to continue on
+//! the parent stack.
+//!
+//! The LR value associated with the parent link is invalid since it points to
+//! the start of the initial function, but this shouldn't block the unwinding
+//! process.
+//!
+//! ## Stack layout
+//!
+//! Note: Non-volatile registers R29, R30, and R31 need to be explicitly
+//! saved/restored on our stack because they are used internally by LLVM and
+//! cannot be used as operands for inline asm (e.g., `inlate()`).
+//!
+//! Here is what the layout of the stack looks like when a coroutine is
+//! suspended.
+//!
+//! ```text
+//!
+//! +--------------+  <- Stack base
+//! | Initial func |  <- Only used once when resuming for the first time.
+//! +--------------+
+//! | Parent link  |  <- The Yielder is a pointer to this address. When the
+//! +--------------+     coroutine is running, it points to the top of the
+//! |              |     parent stack which contains a saved R29, R30, R31,
+//! ~    ...       ~     and LR just like a suspened coroutine.
+//! |              |
+//! +--------------+
+//! |  Saved R31   |
+//! +--------------+
+//! |  Saved R30   |
+//! +--------------+
+//! |  Saved R29   |
+//! +--------------+
+//! |  Saved LR    |
+//! +--------------+
+//! |  Link area   |  <- 32-byte link area
+//! +--------------+
+//! ```
+//!
+//! And this is the layout of the parent stack when a coroutine is running:
+//!
+//! ```text
+//! +--------------+
+//! |     ...      |
+//! |  saved LR    |  <- 32-byte link area.
+//! |     ...      |
+//! +--------------+  <- Stack base.
+//! |  Saved R31   |
+//! +--------------+
+//! |  Saved R30   |
+//! +--------------+
+//! |  Saved R29   |  <- frame pointer chain since FP points to it.
+//! +--------------+  <- Parent link points here.
+//! |  Link area   |  <- 32-byte link area
+//! +--------------+
+//! ```
+//!
+//! And finally, this is the stack layout of a coroutine that has just been
+//! initialized:
+//!
+//! ```text
+//! +--------------+  <- Stack base
+//! | Initial func |
+//! +--------------+
+//! | Parent link  |
+//! +--------------+
+//! |              |
+//! ~ Initial obj  ~
+//! |              |
+//! +--------------+
+//! | Initial PC   |
+//! +--------------+
+//! |  Link area   |  <- 32-byte link area
```

Comment:
No link area is needed here, you can just store all registers in any order (it just needs to match the suspended state).